Address #136: Add support for function annotations

moshez opened the pull request:
Right now this is just a failing test, but hopefully will show what we need to fix.

Diff from pytest:

```
E           <div>                                                               
E           <table class="fieldTable">                                          
E           <tr class="fieldStart">                                             
E           <td class="fieldName">Parameters</td>                               
E           <td class="fieldArg">a</td>
E         - <td>an arg, a the best of args (type: List[str])</td>
E         ?                               ------------------                    
E         + <td>an arg, a the best of args</td>                                 
E           </tr>                                                               
E           <tr class="fieldStart">                                             
E           <td class="fieldName">Returns</td>                                  
E         - <td colspan="2">the best that we can do (type: bool)</td>           
E         ?                                        -------------               
E         + <td colspan="2">the best that we can do</td>
E           </tr>
E           </table>                                                            
E           </div>                                                              
```

mthuurne commented:
This will be useful; thanks!

I'd like the name of the test case to be a bit more specific, for example `test_func_arg_annotation`.

Is `epydoc2stan` the right level to test this at or would it make more sense to test this at the `astbuilder` level? That is, instead of comparing the generated HTML, check the collected information, which is how the variable annotations are tested. I'm a bit worried that when working on for example #121 we might break the test if it depends on the HTML structure. Maybe that is mitigated by the fact we're comparing two HTML forms to each other instead of to a golden result, but I'm still worried that we might be testing too much functionality at once.

Something that also needs a test is how pydoctor should handle the situation where both an annotation and `@type` are present. I think it makes sense to give preference to `@type`, under the assumption that if the author took the effort to manually document the type, there is probably something in there that couldn't be expressed in the type system.

moshez commented:
Hi @mthuurne , thanks for taking the time to comment on a PR that's this raw. This was mostly to explore how we test and how we can implement. I did try to avoid the flattening and epydoc2stan, but it's not completely obvious how to test. As you said, since we are comparing HTML that should be identical, I *think* we are in the clear.

But this is also part of the exploration of the implementation.

This is still very very raw :smile:

moshez commented:
In any case, I have changed the test name more or less to what you asked, and added a test for an epydoc-type override (I agree with your assessment: if the docs took the time to override the mypy annotation, we should use that.)

moshez commented:
Hooray now this is ready for review!

mthuurne commented on pydoctor/epydoc2stan.py:
```diff
@@ -398,6 +400,12 @@ def resolve_types(self):
+        if 'return' not in self.types:
+             return
+        if not self.return_desc:
+            self.return_desc = FieldDesc()
```
Would it be easier to always have a `return_desc` even when it contains no info, or like it is now, where `return_desc` can be `None`? I'm not sure; it probably also depends on what the code looks like after addressing the other issues.

moshez replied:
I cleaned this up by moving the return-special-casing to the named constructor and going through the usual "handle" flow.

mthuurne commented on pydoctor/test/test_epydoc2stan.py:
```diff
@@ -38,6 +40,62 @@ class E:
+        stan = epydoc2stan.format_docstring(docstring)
+        stuff = []
+        flattenString(None, stan).addCallback(stuff.append)
+        return stuff[0].decode('ascii').replace('><', '>\n<')
```
Most of this is already handled by `pydoctor.epydoc.markup.flatten()`, so using that could eliminate some duplicate code.

moshez replied:
Done

moshez commented on pydoctor/astbuilder.py:
```diff
@@ -51,6 +51,32 @@ def node2fullname(expr, ctx):
         return None
 
 
+def _get_all_annotations(func_ast):
```
Done

moshez commented on pydoctor/astbuilder.py:
```diff
@@ -51,6 +51,32 @@ def node2fullname(expr, ctx):
+        if returns:
+            yield 'return', returns
+    return {name: astor.to_source(value).strip() if value is not None else value
+            for name, value in _get_all_ast_annotations(func_ast)}
```
Done

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -292,10 +292,12 @@ def __repr__(self):
 
 class FieldHandler:
-    def __init__(self, obj):
+    def __init__(self, obj, annotations=None):
```
Done

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -292,10 +292,12 @@ def __repr__(self):
         self.obj = obj
 
         self.types = {}
+        if annotations is not None:
```
Done

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -398,6 +400,12 @@ def resolve_types(self):
             if pd.name in self.types:
                 pd.type = self.types[pd.name]
+        if 'return' not in self.types:
+             return
```
Completely agreed with everything here. Opened https://github.com/twisted/pydoctor/issues/216

moshez commented:
Thanks @mthuurne for the review. I have addressed most comments. There were two things I only partially addressed, because fully addressing them felt like adding unrelated features:

* Adding annotations to new code: I did add an argument to the `FieldHandler` constructor, so technically this function definition is "new code". However, adding annotations to the constructor meant annotating all object attributes, and that seems like it would be better be done by moving the object to be a dataclass or an attrs class, which definitely did not feel like a part of this effort.
* Documenting things which only have a type -- I have addressed the specific request you had (documenting the return type when only a type is provided) since this was easy enough and matched existing pydoctor behavior. Adding the ability to document type-only parameters, again, felt like this was a different issue. If this makes sense to you, I'm happy to open the issue. 

All other things you have noted should have been addressed.

moshez commented on pydoctor/astbuilder.py:
```diff
@@ -51,6 +52,33 @@ def node2fullname(expr, ctx):
         return None
 
 
+def _get_all_annotations(func_ast: ast.FunctionDef) -> Mapping[str, ast.expr]:
```
done.

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -307,6 +311,19 @@ def __init__(self, obj):
         self.unattached_types = {}
 
+    @classmethod
+    def from_ast_annotations(cls, obj: Any, annotations: Mapping[str, ast.expr]) -> "FieldHandler":
```
done

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -493,12 +510,11 @@ def format_docstring(obj):
-            fh.handle(Field(field, obj))
-        fh.resolve_types()
-        s(fh.format())
+    fh = FieldHandler.from_ast_annotations(obj, getattr(source, 'annotations', {}))
```
sounds good.

moshez commented on pydoctor/epydoc2stan.py:
```diff
@@ -307,6 +311,19 @@ def __init__(self, obj):
 
+    @classmethod
+    def from_ast_annotations(cls, obj: Any, annotations: Mapping[str, ast.expr]) -> "FieldHandler":
+        annotations = {name: AnnotationDocstring(value).to_stan(_EpydocLinker(obj))
```
done

moshez commented:
Hi @mthuurne I've addressed the issues you pointed out. Since I do not have permission to merge, I would appreciate it if you could merge it. Thanks in advance!
